### PR TITLE
fix(gui-client): defer GUI exit until tunnel closes

### LINKED ISSUE #6873: bug(rust/gui-client/windows): delay when quitting related to Sentry
#### Describe the bug

The application takes 12 seconds to exit when clicking "Quit" in the menu.

It seems to only happen if Firezone is signed in, and if we're on Windows. Windows is known to be a little slower for tunnel start / stop than Linux.

Hypotheses:

- Inherent to Sentry? - No, it only replicates when Firezone is signed in
- Related to Windows? - Possibly, haven't seen it replicate on Linux
- Sentry is just slow in debug builds - No, happens in release build, and doesn't happen in Linux debug builds
- Related to GUI / tunnel daemon process split? - Maybe. Doesn't seem to happen in Headless Client.
- Related to DNS? - Likely.
- Related to Internet Resource? - Unlikely, replicates even if the Internet Resource is disabled.

#### To Reproduce

1. Start the tunnel daemon (IPC service)
2. Start the GUI
3. Let it sign in
4. Click "Quit" in the menu
5. There's a 5-second delay while trying to flush telemetry
6. There's an extra delay before exiting

#### Expected behavior

Expect the app to quit instantly if there are no errors to send

#### Screenshots / Logs

Windows (bad):

image

Linux (good):

image

#### Platform (please complete the following information)

- _Component (i.e. macOS client / Linux client / Gateway / Admin portal):_ GUI Client GUI process
- _Firezone Version (e.g. 1.0.0 or N/A):_ d2a8155ba (tip of `main`)
- _OS and version: (e.g. Ubuntu 22.04 or N/A):_ Windows
- _Deployment method: (e.g. Docker / Systemd / App Store or N/A):_ Built from source

### PR DESCRIPTION
**Pull request by ReactorScram**
Closes #6873

The issue seems to be a race between flushing Sentry in the GUI process and shutting down Firezone in the tunnel daemon (IPC service).

With this change, the GUI waits to hear `DisconnectedGracefully` from the tunnel daemon before flushing Sentry, and the issue is prevented.

Adding the new state and new IPC message required small changes in several places

### CODE REVIEW

**Comment by ReactorScram** [comment]
Is it worth adding testing around `Controller` here? This is sort of a weird corner case... If we tested `Controller` without running the tunnel, it would not have triggered

File: rust/gui-client/src-common/src/controller.rs [hunk 1]
```diff
@@ -142,8 +142,11 @@ pub enum Status {
         /// The token to log in to the Portal, for retrying the connection request.
         token: SecretString,
     },
+    Quitting, // The user asked to quit and we're waiting for the tunnel daemon to gracefully disconnect so we can flush telemetry.
     /// Firezone is ready to use.
-    TunnelReady { resources: Vec<ResourceDescription> },
+    TunnelReady {
+        resources: Vec<ResourceDescription>,
+    },
```
Comment (ReactorScram) [inline rust/gui-client/src-common/src/controller.rs:149]: Not sure why `cargo fmt` changed its mind on this

File: rust/gui-client/src-common/src/controller.rs [hunk 2]
```diff
@@ -165,16 +168,28 @@ impl Default for Status {
...
             Status::Disconnected | Status::RetryingConnection { .. } => false,
+            Status::Quitting => false,
             Status::TunnelReady { .. }
             | Status::WaitingForPortal { .. }
             | Status::WaitingForTunnel { .. } => true,
         }
     }
 
+    /// True if we should react to `OnUpdateResources`
+    fn needs_resource_updates(&self) -> bool {
```
Comment (ReactorScram) [inline rust/gui-client/src-common/src/controller.rs:183]: Unrelated refactor but wasn't quite big enough to split out

File: website/src/components/Changelog/GUI.tsx [hunk 3]
```diff
@@ -26,6 +26,9 @@ export default function GUI({ title }: { title: string }) {
         <ChangeItem pull="6782">
           Adds always-on error reporting using sentry.io.
         </ChangeItem>
+        <ChangeItem enable={title === "Windows"} pull="6874">
+          Fixes a delay when closing the GUI
+        </ChangeItem>
```
Comment (ReactorScram) [inline website/src/components/Changelog/GUI.tsx]: I couldn't replicate it on the Headless Client, this is explained by the Headless Client not having the IPC boundary, so it's just one thread and doesn't race between shutting down connlib and shutting down Sentry.

All I can think for the Linux GUI Client is that Linux shuts down a lot quicker, so there is a race but it never happened to trigger for me.

File: rust/gui-client/src-common/src/controller.rs [hunk 4]
```diff
@@ -245,9 +260,12 @@ impl<I: GuiIntegration> Controller<I> {
                     }
                     self.try_retry_connection().await?
                 }
-                event = self.ipc_rx.recv() => self.handle_ipc_event(event.context("IPC task stopped")?).await?,
+                event = self.ipc_rx.recv() => if let ControlFlow::Break(()) = self.handle_ipc_event(event.context("IPC task stopped")?).await? {
+                    break;
+                },
```
Comment (thomaseizinger) [inline rust/gui-client/src-common/src/controller.rs]: A lot of control-flow going on here. Should we split this over a few lines perhaps?

File: rust/gui-client/src-common/src/controller.rs [hunk 5]
```diff
@@ -487,9 +510,18 @@ impl<I: GuiIntegration> Controller<I> {
...
                     Error::Other(anyhow!("Couldn't send `ClearLogs` result to Tauri task"))
                 })?;
-                Ok(())
             }
-            IpcServerMsg::ConnectResult(result) => self.handle_connect_result(result).await,
+            IpcServerMsg::ConnectResult(result) => {
+                return self
+                    .handle_connect_result(result)
+                    .await
+                    .map(|_| ControlFlow::Continue(()))
```
Comment (thomaseizinger) [inline rust/gui-client/src-common/src/controller.rs:521]: Would prefer use of `?`.

File: rust/headless-client/src/main.rs [hunk 6]
```diff
@@ -318,7 +318,11 @@ fn main() -> Result<()> {
         session.disconnect();
 
         result
-    })
+    });
+
+    tracing::warn!("Stopping telemetry...");
```
Comment (thomaseizinger) [inline rust/headless-client/src/main.rs]: Why warn?

File: website/src/components/Changelog/GUI.tsx [hunk 7]
```diff
@@ -26,6 +26,9 @@ export default function GUI({ title }: { title: string }) {
         <ChangeItem pull="6782">
           Adds always-on error reporting using sentry.io.
         </ChangeItem>
+        <ChangeItem enable={title === "Windows"} pull="6874">
+          Fixes a delay when closing the GUI
```
Comment (thomaseizinger) [inline website/src/components/Changelog/GUI.tsx]:
```suggestion
          Fixes a delay when closing the GUI.
```

File: rust/gui-client/src-common/src/controller.rs [hunk 8]
```diff
@@ -712,12 +748,20 @@ impl<I: GuiIntegration> Controller<I> {
 
     /// Deletes the auth token, stops connlib, and refreshes the tray menu
     async fn sign_out(&mut self) -> Result<()> {
+        match self.status {
+            Status::Disconnected
+            | Status::RetryingConnection { .. }
+            | Status::TunnelReady { .. }
+            | Status::WaitingForPortal { .. }
+            | Status::WaitingForTunnel { .. } => {}
+            Status::Quitting => return Ok(()),
```
Comment (thomaseizinger) [inline rust/gui-client/src-common/src/controller.rs]: Can we have this branch at the top? It wasn't immediately obvious to me that this one bails early.

**Comment by thomaseizinger** [comment]
I slightly reworded the title, hope you don't mind :)

**Review by thomaseizinger — APPROVED** [review]

**Comment by ReactorScram** [comment]
I reworded it again, we aren't actually waiting for the IPC service itself to stop, we're waiting for the tunnel to close down

File: rust/gui-client/src-common/src/controller.rs (hunk 5 above)
Comment (ReactorScram) [reply rust/gui-client/src-common/src/controller.rs:521]: I tried a couple different things, including `return Ok(ControlFlow::from(self.handle_connect_result(result).await?));`, and none of them compiled?
